File: retiro-del-rocio/app/Models/ChatMessage.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

/**
 * One message in a guest's Concierge Chat thread with the front desk —
 * scoped to a booking, so a new guest checking into the same room never
 * sees the previous occupant's conversation.
 */
class ChatMessage extends Model
{
    public const GUEST = 'guest';

    public const STAFF = 'staff';

    protected $fillable = [
        'booking_id', 'sender_type', 'sender_name', 'body', 'read_at',
    ];

    protected $casts = [
        'read_at' => 'datetime',
    ];

    public function booking()
    {
        return $this->belongsTo(Booking::class);
    }

    public function isFromGuest(): bool
    {
        return $this->sender_type === self::GUEST;
    }

    /** The payload the guest tablet renders in the Concierge Chat thread. */
    public function toGuestArray(): array
    {
        return [
            'id' => $this->id,
            'sender_type' => $this->sender_type,
            'sender_name' => $this->sender_name,
            'body' => $this->body,
            'is_mine' => $this->isFromGuest(),
            'created_at' => $this->created_at?->toIso8601String(),
            'time_label' => optional($this->created_at)->format('g:i A'),
        ];
    }

    /** The same thread, from reception's side — `is_mine` flips to the staff reply. */
    public function toStaffArray(): array
    {
        return [
            ...$this->toGuestArray(),
            'is_mine' => ! $this->isFromGuest(),
        ];
    }
}

File: retirodelrocioapp/lib/features/guest/chat/domain/chat_message.dart
import 'package:flutter/foundation.dart';

/// Who sent a [ChatMessage] — the guest, or a staff member replying from the
/// front desk.
enum ChatSender {
  guest,
  staff;

  static ChatSender fromApi(String? value) =>
      value == 'staff' ? ChatSender.staff : ChatSender.guest;
}

/// One message in the guest's Concierge Chat thread with the front desk
/// (`GET /chat/messages`).
@immutable
class ChatMessage {
  const ChatMessage({
    required this.id,
    required this.sender,
    required this.senderName,
    required this.body,
    required this.isMine,
    required this.timeLabel,
    required this.createdAt,
  });

  final int id;
  final ChatSender sender;
  final String? senderName;
  final String body;

  /// True for a message the guest themself sent — the right-aligned, gold
  /// bubble; false renders left-aligned as the front desk's reply.
  final bool isMine;

  final String timeLabel;
  final DateTime createdAt;

  factory ChatMessage.fromJson(Map<String, dynamic> json) => ChatMessage(
    id: (json['id'] as num?)?.toInt() ?? 0,
    sender: ChatSender.fromApi(json['sender_type'] as String?),
    senderName: json['sender_name'] as String?,
    body: json['body'] as String? ?? '',
    isMine: json['is_mine'] as bool? ?? false,
    timeLabel: json['time_label'] as String? ?? '',
    createdAt:
        DateTime.tryParse(json['created_at'] as String? ?? '') ??
        DateTime.now(),
  );
}
